File: model-client/src/main/java/org/modelix/model/lazy/ObjectStoreCache.java
package org.modelix.model.lazy;

import org.apache.commons.collections4.map.LRUMap;
import org.modelix.model.IKeyValueStore;
import org.modelix.model.util.StreamUtils;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.BiFunction;
import java.util.function.Function;
import java.util.stream.Collectors;

public class ObjectStoreCache implements IDeserializingKeyValueStore {
  private static final Object NULL = new Object();

  private IKeyValueStore store;
  private final Map<String, Object> cache = Collections.synchronizedMap(new LRUMap<>(100000));

  public ObjectStoreCache(IKeyValueStore store1) {
    store = store1;
  }

  @Override
  public IKeyValueStore getKeyValueStore() {
    return store;
  }

  @Override
  public <T> Iterable<T> getAll(Iterable<String> hashes_, BiFunction<String, String, T> deserializer) {
    List<String> hashes = StreamUtils.INSTANCE.toStream(hashes_).collect(Collectors.toList());
    final Map<String, T> result = new HashMap<>();
    List<String> nonCachedHashes = new ArrayList<>(hashes.size());

    for (String hash : hashes) {
      T deserialized = (T) cache.get(hash);
      if (deserialized == null) {
        nonCachedHashes.add(hash);
      } else {
        result.put(hash, (deserialized == NULL ? null : deserialized));
      }
    }

    if (!nonCachedHashes.isEmpty()) {
      for (Map.Entry<String, String> entry : store.getAll(nonCachedHashes).entrySet()) {
        String hash = entry.getKey();
        String serialized = entry.getValue();
        if (serialized == null) {
          result.put(hash, null);
        } else {
          T deserialized = deserializer.apply(hash, serialized);
          cache.put(hash, (deserialized == null ? NULL : deserialized));
          result.put(hash, deserialized);
        }
      }
    }

    return hashes.stream().map(result::get)::iterator;
  }

  public <T> T get(String hash, Function<String, T> deserializer) {
    if (hash == null) {
      return null;
    }
    T deserialized = (T) cache.get(hash);
    if (deserialized == null) {
      String serialized = store.get(hash);
      if (serialized == null) {
        return null;
      }
      deserialized = deserializer.apply(serialized);
      cache.put(hash, (deserialized == null ? NULL : deserialized));
    }
    return (deserialized == NULL ? null : deserialized);
  }

  public void put(String hash, Object deserialized, String serialized) {
    store.put(hash, serialized);
    cache.put(hash, (deserialized == null ? NULL : deserialized));
  }

  public void clearCache() {
    cache.clear();
  }

  @Override
  public void prefetch(String hash) {
    store.prefetch(hash);
  }
}

File: model-client/src/main/kotlin/org/modelix/model/lazy/ObjectStoreCache.kt
package org.modelix.model.lazy

import org.apache.commons.collections4.map.LRUMap
import org.modelix.model.IKeyValueStore
import org.modelix.model.util.StreamUtils.toStream
import java.util.*
import java.util.function.BiFunction
import java.util.function.Function
import java.util.stream.Collectors

class ObjectStoreCache(override val keyValueStore: IKeyValueStore) : IDeserializingKeyValueStore {
    private val cache = Collections.synchronizedMap(LRUMap<String?, Any>(100000))

    override fun <T> getAll(hashes_: Iterable<String?>?, deserializer: BiFunction<String?, String?, T>?): Iterable<T>? {
        val hashes = toStream(hashes_!!).collect(Collectors.toList())
        val result: MutableMap<String?, T?> = HashMap()
        val nonCachedHashes: MutableList<String?> = ArrayList(hashes.size)
        for (hash in hashes) {
            val deserialized = cache[hash] as T?
            if (deserialized == null) {
                nonCachedHashes.add(hash)
            } else {
                result[hash] = if (deserialized === NULL) null else deserialized
            }
        }
        if (!nonCachedHashes.isEmpty()) {
            for ((hash, serialized) in keyValueStore.getAll(nonCachedHashes)!!) {
                if (serialized == null) {
                    result[hash] = null
                } else {
                    val deserialized: T? = deserializer!!.apply(hash, serialized)
                    cache[hash] = deserialized ?: NULL
                    result[hash] = deserialized
                }
            }
        }
        return Iterable<T> { hashes.stream().map { key: String? -> result[key] }.iterator() as Iterator<T> }
    }

    override fun <T> get(hash: String?, deserializer: Function<String?, T>?): T? {
        if (hash == null) {
            return null
        }
        var deserialized = cache[hash] as T?
        if (deserialized == null) {
            val serialized = keyValueStore[hash] ?: return null
            deserialized = deserializer!!.apply(serialized)
            cache[hash] = deserialized ?: NULL
        }
        return if (deserialized === NULL) null else deserialized
    }

    override fun put(hash: String?, deserialized: Any?, serialized: String?) {
        keyValueStore.put(hash, serialized)
        cache[hash] = deserialized ?: NULL
    }

    fun clearCache() {
        cache.clear()
    }

    override fun prefetch(hash: String?) {
        keyValueStore.prefetch(hash)
    }

    companion object {
        private val NULL = Any()
    }

}